### PR TITLE
homework 39 solution

### PR DESCRIPTION
**Pull request by tkhambir**

### CODE REVIEW

File: src/main/java/com/dev/cinema/model/User.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        this.password = password;
+    }
+
+    @Override
+    public String toString() {
+        return "User{"
+                + "id=" + id
+                + ", email='" + email + '\''
+                + ", password='" + password + '\''
+                + '}';
```
Comment (dariiabutorina) [inline src/main/java/com/dev/cinema/model/User.java]: Let's not show the password

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Override
+    public User register(String email, String password) {
+        User user = new User();
+        byte[] salt = HashUtil.getSalt();
+        String hashPassword = HashUtil.hashPassword(password, salt);
+        user.setEmail(email);
+        user.setPassword(hashPassword);
+        user.setSalt(salt);
+        return userService.add(user);
```
Comment (dariiabutorina) [inline src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]: All the business logic somehow related to initializing or setting the values to our fields must be on the **_service_** layer.
If somebody creates the object **_User_** not throughout the **_register_** method but using the **_userService.create()_** instead, we'll not have its password hashed, which might cause some mistakes in the future while working with the object or with the security itself.
So let's move the hashing logic to the **_service_** layer

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    private static boolean isValid(String password, User user) {
```
Comment (dariiabutorina) [inline src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]: Why static?

File: src/main/java/com/dev/cinema/util/HashUtil.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        SecureRandom random = new SecureRandom();
+        byte[] salt = new byte[16];
+        random.nextBytes(salt);
+        return salt;
+    }
+
+    public static String hashPassword(String password, byte[] salt) {
+        StringBuilder hashedPassword = new StringBuilder();
+
+        try {
```
Comment (dariiabutorina) [inline src/main/java/com/dev/cinema/util/HashUtil.java:20]:
```suggestion
        StringBuilder hashedPassword = new StringBuilder();
        try {
```

File: src/main/java/com/dev/cinema/util/HashUtil.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            MessageDigest messageDigest = MessageDigest.getInstance(HASH_ALGORITHM);
+            messageDigest.update(salt);
+            byte[] digest = messageDigest.digest(password.getBytes());
+            for (byte b : digest) {
+                hashedPassword.append(String.format("%02x", b));
+            }
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Can't hash password", e);
```
Comment (dariiabutorina) [inline src/main/java/com/dev/cinema/util/HashUtil.java:28]: The cause - ?

**Review by dariiabutorina — CHANGES_REQUESTED** [review]

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java (hunk 2 above)
Comment (RomanDubovskyi) [reply src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]: > All the business logic somehow related to initializing or setting the values to our fields must be on the **_service_** layer.
> If somebody creates the object **_User_** not throughout the **_register_** method but using the **_userService.create()_** instead, we'll not have its password hashed, which might cause some mistakes in the future while working with the object or with the security itself.
> So let's move the hashing logic to the **_service_** layer

\+

File: src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java (hunk 3 above)
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]:
```suggestion
    private static boolean isPasswordValid(String password, User user) {
```

Comment (RomanDubovskyi) [reply src/main/java/com/dev/cinema/security/AuthenticationServiceImpl.java]: > Why static?

\+

File: src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+@Service
+public class UserServiceImpl implements UserService {
+    @Inject
+    private UserDao userDao;
+
+    @Override
+    public User add(User user) {
+        byte[] salt = HashUtil.getSalt();
+        user.setSalt(salt);
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/service/impl/UserServiceImpl.java:19]:
```suggestion
        user.setSalt(HashUtil.getSalt());
```

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**